pex/pex_bootstrapper.py

```diff
@@ -160,39 +160,68 @@ def _select_path_interpreter(
     path=None,  # type: Optional[str]
     valid_basenames=None,  # type: Optional[Tuple[str, ...]]
     interpreter_constraints=None,  # type: Optional[Iterable[str]]
+    preferred_interpreter=None,  # type: Optional[PythonInterpreter]
 ):
     # type: (...) -> Optional[PythonInterpreter]
     candidate_interpreters_iter = iter_compatible_interpreters(
         path=path,
         valid_basenames=valid_basenames,
         interpreter_constraints=interpreter_constraints,
+        preferred_interpreter=preferred_interpreter,
     )
     current_interpreter = PythonInterpreter.get()  # type: PythonInterpreter
-    candidate_interpreters = []
+    preferred_interpreter = preferred_interpreter or current_interpreter
+    candidate_interpreters = OrderedSet()  # type: OrderedSet[PythonInterpreter]
     for interpreter in candidate_interpreters_iter:
-        if current_interpreter == interpreter:
-            # Always prefer continuing with the current interpreter when possible to avoid re-exec
-            # overhead.
-            return current_interpreter
+        if preferred_interpreter == interpreter:
+            # Always respect the preferred interpreter if it doesn't violate other constraints.
+            return preferred_interpreter
         else:
-            candidate_interpreters.append(interpreter)
+            candidate_interpreters.add(interpreter)
     if not candidate_interpreters:
         return None
-
+    if current_interpreter in candidate_interpreters:
+        # Always prefer continuing with the current interpreter when possible to avoid re-exec
+        # overhead.
+        return current_interpreter
     # TODO: Allow the selection strategy to be parameterized:
     #   https://github.com/pantsbuild/pex/issues/430
     return PythonInterpreter.latest_release_of_min_compatible_version(candidate_interpreters)
 
 
 def find_compatible_interpreter(interpreter_constraints=None):
     # type: (Optional[Iterable[str]]) -> PythonInterpreter
+
+    def gather_constraints():
+        # type: () -> Iterable[str]
+        constraints = []
+        if ENV.PEX_PYTHON:
+            constraints.append("PEX_PYTHON={}".format(ENV.PEX_PYTHON))
+        if ENV.PEX_PYTHON_PATH:
+            constraints.append("PEX_PYTHON_PATH={}".format(ENV.PEX_PYTHON_PATH))
+        if interpreter_constraints:
+            constraints.append("Version matches {}".format(" or ".join(interpreter_constraints)))
+        return constraints
+
+    preferred_interpreter = None  # type: Optional[PythonInterpreter]
+    if ENV.PEX_PYTHON and os.path.isabs(ENV.PEX_PYTHON):
+        try:
+            preferred_interpreter = PythonInterpreter.from_binary(ENV.PEX_PYTHON)
```

Comment:
As mentioned above, should this immediately error if there are `interpreter_constraints` which don't match?

Reply:
Whoops: stale. It does error slightly below.